File: solutions/code-carl/ch05_stack/t4_lc1047_remove_duplicates.py
"""LeetCode 1047. Remove All Adjacent Duplicates In String
https://leetcode.cn/problems/remove-all-adjacent-duplicates-in-string/
"""


class Solution:
    def removeDuplicates(self, s: str) -> str:
        chars = list(s)
        slow = 0
        for fast in range(len(chars)):
            if slow > 0 and chars[fast] == chars[slow - 1]:
                slow -= 1
            else:
                chars[slow] = chars[fast]
                slow += 1
        return "".join(chars[0:slow])


if __name__ == "__main__":
    solution = Solution()
    assert solution.removeDuplicates("abbaca") == "ca"
    assert solution.removeDuplicates("azxxzy") == "ay"
    print("All tests passed.")

File: solutions/java/code-carl/ch05_stack/T4_LC1047_RemoveDuplicates.java
/**
 * Practice template for:
 * - LeetCode 1047. Remove All Adjacent Duplicates In String
 *   https://leetcode.cn/problems/remove-all-adjacent-duplicates-in-string/
 */
public class T4_LC1047_RemoveDuplicates {
    public String removeDuplicates(String s) {
        char[] chars = s.toCharArray();
        int slow = 0, fast = 0;
        for (fast = 0;  fast < s.length(); fast++) {
            if (slow > 0 && chars[fast] == chars[slow - 1]) {
                slow--;
            } else {
                chars[slow] = chars[fast];
                slow++;
            }
        }
        return new String(chars, 0, slow);
    }
}
